tool/common/update/update.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	return "", false, nil
+}
+
+func Download(toolsVersion string) error {
+	// If the version of the running binary or the version downloaded to
+	// $TELEPORT_HOME/bin is the same as the requested version of client tools,
+	// nothing to be done, exit early.
+	teleportVersion, err := version()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+	if toolsVersion == teleport.Version || toolsVersion == teleportVersion {
+		return nil
+	}
+
+	// Create $TELEPORT_HOME/bin if it does not exist.
+	dir, err := toolsDir()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Download and update {tsh, tctl} in $TELEPORT_HOME/bin.
+	if err := update(toolsVersion); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// TODO(russjones): Add edition check here as well.
+func update(toolsVersion string) error {
+	dir, err := toolsDir()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Lock to allow multiple concurrent {tsh, tctl} to run.
+	unlock, err := lock(dir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer unlock()
+
+	// Get platform specific download URLs.
+	archiveURL, hashURL, err := urls(toolsVersion, "")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	log.Debugf("Archive download path: %v.", archiveURL)
+
+	// Download the archive and validate against the hash. Download to a
+	// temporary path within $TELEPORT_HOME/bin.
+	hash, err := downloadHash(hashURL)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	path, err := downloadArchive(archiveURL, hash)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer os.Remove(path)
+
+	// Perform atomic replace so concurrent exec do not fail.
+	if err := replace(path, hash); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// urls returns the URL for the Teleport archive to download. The format is:
+// https://cdn.teleport.dev/teleport-{, ent-}v15.3.0-{linux, darwin, windows}-{amd64,arm64,arm,386}-{fips-}bin.tar.gz
+func urls(toolsVersion string, toolsEdition string) (string, string, error) {
+	var archive string
+
+	switch runtime.GOOS {
+	case "darwin":
+		archive = baseUrl + "/tsh-" + toolsVersion + ".pkg"
+	case "windows":
+		archive = baseUrl + "/teleport-v" + toolsVersion + "-windows-amd64-bin.zip"
+	case "linux":
+		edition := ""
+		if toolsEdition == "ent" || toolsEdition == "fips" {
+			edition = "ent-"
+		}
+		fips := ""
+		if toolsEdition == "fips" {
+			fips = "fips-"
+		}
+
+		var b strings.Builder
+		b.WriteString(baseUrl + "/teleport-")
+		if edition != "" {
+			b.WriteString(edition)
+		}
+		b.WriteString("v" + toolsVersion + "-" + runtime.GOOS + "-" + runtime.GOARCH + "-")
+		if fips != "" {
+			b.WriteString(fips)
+		}
+		b.WriteString("bin.tar.gz")
+		archive = b.String()
+	default:
+		return "", "", trace.BadParameter("unsupported runtime: %v", runtime.GOOS)
+	}
+
+	return archive, archive + ".sha256", nil
+}
+
```

Comment:
I put together a more productionized version of these methods, which adds disk space checking, timeouts, context cancellation, etc.: https://github.com/gravitational/teleport/compare/sclevine/teleport-update-enable

We could try to reuse those, but the use cases are a bit different. For now, you might want to look at some of the differences. E.g., use of the client and contexts to handle graceful cancellation, and limiting the number of bytes read over the network.